File: native/android/MediaModule.kt
/**
 * MediaModule - Android TurboModule for camera capture and photo library access
 *
 * Provides:
 * - captureImage: Launches system camera via TakePicture contract
 * - pickFromLibrary: Launches Android Photo Picker (API 33+) or system file picker fallback
 * - captureFromView: Snapshots an inline SDKCameraView by its cameraId
 * - Permission checking/requesting for camera and photo library
 *
 * Options and results are JSON-stringified for Codegen compatibility.
 * Matches TurboModule spec: NativeMediaModule.ts
 * Registered name: "MediaModule"
 */

package com.anthropic.sdk.media

import android.Manifest
import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import android.os.Build
import android.provider.MediaStore
import android.util.Base64
import androidx.activity.result.contract.ActivityResultContracts
import androidx.core.content.ContextCompat
import androidx.core.content.FileProvider
import com.facebook.react.bridge.ActivityEventListener
import com.facebook.react.bridge.BaseActivityEventListener
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import org.json.JSONArray
import org.json.JSONObject
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.FileOutputStream

class MediaModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    companion object {
        const val NAME = "MediaModule"
        private const val REQUEST_CAMERA = 1001
        private const val REQUEST_LIBRARY = 1002
        private const val REQUEST_CAMERA_PERMISSION = 2001
        private const val REQUEST_LIBRARY_PERMISSION = 2002
    }

    override fun getName(): String = NAME

    private var pendingPromise: Promise? = null
    private var pendingOptions: JSONObject? = null
    private var captureUri: Uri? = null

    init {
        val listener: ActivityEventListener = object : BaseActivityEventListener() {
            override fun onActivityResult(activity: Activity, requestCode: Int, resultCode: Int, data: Intent?) {
                when (requestCode) {
                    REQUEST_CAMERA -> handleCameraResult(resultCode, data)
                    REQUEST_LIBRARY -> handleLibraryResult(resultCode, data)
                }
            }
        }
        reactContext.addActivityEventListener(listener)
    }

    // -----------------------------------------------------------------------
    // captureImage
    // -----------------------------------------------------------------------

    @ReactMethod
    fun captureImage(optionsStr: String, promise: Promise) {
        try {
            val options = JSONObject(optionsStr)
            pendingPromise = promise
            pendingOptions = options

            val activity = currentActivity
            if (activity == null) {
                promise.reject("MEDIA_ERROR", "No current activity")
                return
            }

            // Create temp file for camera output
            val tempDir = reactApplicationContext.cacheDir
            val tempFile = File.createTempFile("capture_", ".jpg", tempDir)
            captureUri = FileProvider.getUriForFile(
                reactApplicationContext,
                "${reactApplicationContext.packageName}.fileprovider",
                tempFile
            )

            val intent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
            intent.putExtra(MediaStore.EXTRA_OUTPUT, captureUri)
            intent.addFlags(Intent.FLAG_GRANT_WRITE_URI_PERMISSION)
            activity.startActivityForResult(intent, REQUEST_CAMERA)
        } catch (e: Exception) {
            promise.reject("MEDIA_ERROR", "Failed to launch camera: ${e.message}", e)
        }
    }

    // -----------------------------------------------------------------------
    // pickFromLibrary
    // -----------------------------------------------------------------------

    @ReactMethod
    fun pickFromLibrary(optionsStr: String, promise: Promise) {
        try {
            val options = JSONObject(optionsStr)
            pendingPromise = promise
            pendingOptions = options

            val activity = currentActivity
            if (activity == null) {
                promise.reject("MEDIA_ERROR", "No current activity")
                return
            }

            val multiple = options.optBoolean("multiple", false)

            val intent = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                // Android 13+ Photo Picker
                Intent(MediaStore.ACTION_PICK_IMAGES).apply {
                    type = options.optString("accept", "image/*")
                    if (multiple) {
                        val maxCount = options.optInt("maxCount", 10)
                        putExtra(MediaStore.EXTRA_PICK_IMAGES_MAX, maxCount)
                    }
                }
            } else {
                Intent(Intent.ACTION_GET_CONTENT).apply {
                    type = options.optString("accept", "image/*")
                    putExtra(Intent.EXTRA_ALLOW_MULTIPLE, multiple)
                }
            }

            activity.startActivityForResult(intent, REQUEST_LIBRARY)
        } catch (e: Exception) {
            promise.reject("MEDIA_ERROR", "Failed to launch photo picker: ${e.message}", e)
        }
    }

    // -----------------------------------------------------------------------
    // captureFromView
    // -----------------------------------------------------------------------

    @ReactMethod
    fun captureFromView(cameraId: String, optionsStr: String, promise: Promise) {
        try {
            if (cameraId.isEmpty()) {
                promise.reject("MEDIA_ERROR", "captureFromView requires a non-empty cameraId")
                return
            }

            val options = JSONObject(optionsStr)
            val view = CameraViewManager.getView(cameraId)

            if (view == null) {
                promise.reject("MEDIA_ERROR", "Camera view with id '$cameraId' not found")
                return
            }

            view.captureFrame { bitmap, error ->
                if (error != null || bitmap == null) {
                    promise.reject("MEDIA_ERROR", "Failed to capture frame: ${error?.message ?: "unknown"}")
                    return@captureFrame
                }

                val quality = (options.optDouble("quality", 0.8) * 100).toInt()
                val maxDimension = if (options.has("maxDimension")) options.optInt("maxDimension") else null
                val includeBase64 = options.optBoolean("includeBase64", false)

                val resized = resizeIfNeeded(bitmap, maxDimension)
                val result = buildMediaResult(resized, quality, includeBase64, "viewCapture")
                promise.resolve(result)
            }
        } catch (e: Exception) {
            promise.reject("MEDIA_ERROR", "captureFromView failed: ${e.message}", e)
        }
    }

    // -----------------------------------------------------------------------
    // Permissions
    // -----------------------------------------------------------------------

    @ReactMethod
    fun checkCameraPermission(promise: Promise) {
        val status = ContextCompat.checkSelfPermission(
            reactApplicationContext,
            Manifest.permission.CAMERA
        )
        promise.resolve(if (status == PackageManager.PERMISSION_GRANTED) "granted" else "denied")
    }

    @ReactMethod
    fun checkLibraryPermission(promise: Promise) {
        val permission = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            Manifest.permission.READ_MEDIA_IMAGES
        } else {
            Manifest.permission.READ_EXTERNAL_STORAGE
        }
        val status = ContextCompat.checkSelfPermission(reactApplicationContext, permission)
        promise.resolve(if (status == PackageManager.PERMISSION_GRANTED) "granted" else "denied")
    }

    @ReactMethod
    fun requestCameraPermission(promise: Promise) {
        val status = ContextCompat.checkSelfPermission(
            reactApplicationContext,
            Manifest.permission.CAMERA
        )
        if (status == PackageManager.PERMISSION_GRANTED) {
            promise.resolve("granted")
        } else {
            // In production this would use ActivityCompat.requestPermissions
            // with a callback. Simplified for the module spec.
            promise.resolve("denied")
        }
    }

    @ReactMethod
    fun requestLibraryPermission(promise: Promise) {
        val permission = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            Manifest.permission.READ_MEDIA_IMAGES
        } else {
            Manifest.permission.READ_EXTERNAL_STORAGE
        }
        val status = ContextCompat.checkSelfPermission(reactApplicationContext, permission)
        if (status == PackageManager.PERMISSION_GRANTED) {
            promise.resolve("granted")
        } else {
            promise.resolve("denied")
        }
    }

    // -----------------------------------------------------------------------
    // Activity Result Handlers
    // -----------------------------------------------------------------------

    private fun handleCameraResult(resultCode: Int, data: Intent?) {
        val promise = pendingPromise ?: return
        val options = pendingOptions ?: JSONObject()

        if (resultCode != Activity.RESULT_OK) {
            promise.reject("MEDIA_CANCELLED", "User cancelled camera capture")
            clearPending()
            return
        }

        try {
            val uri = captureUri ?: return
            val stream = reactApplicationContext.contentResolver.openInputStream(uri)
            val bitmap = BitmapFactory.decodeStream(stream)
            stream?.close()

            if (bitmap == null) {
                promise.reject("MEDIA_ERROR", "Failed to decode captured image")
                clearPending()
                return
            }

            val quality = (options.optDouble("quality", 0.8) * 100).toInt()
            val maxDimension = if (options.has("maxDimension")) options.optInt("maxDimension") else null
            val includeBase64 = options.optBoolean("includeBase64", false)

            val resized = resizeIfNeeded(bitmap, maxDimension)
            val result = buildMediaResult(resized, quality, includeBase64, "capture")
            promise.resolve(result)
        } catch (e: Exception) {
            promise.reject("MEDIA_ERROR", "Failed to process camera result: ${e.message}", e)
        }

        clearPending()
    }

    private fun handleLibraryResult(resultCode: Int, data: Intent?) {
        val promise = pendingPromise ?: return
        val options = pendingOptions ?: JSONObject()

        if (resultCode != Activity.RESULT_OK || data == null) {
            promise.reject("MEDIA_CANCELLED", "User cancelled photo selection")
            clearPending()
            return
        }

        try {
            val quality = (options.optDouble("quality", 0.8) * 100).toInt()
            val maxDimension = if (options.has("maxDimension")) options.optInt("maxDimension") else null
            val includeBase64 = options.optBoolean("includeBase64", false)
            val multiple = options.optBoolean("multiple", false)

            val uris = mutableListOf<Uri>()

            // Check for multiple selections
            if (multiple && data.clipData != null) {
                val clipData = data.clipData!!
                val maxCount = options.optInt("maxCount", 10)
                for (i in 0 until minOf(clipData.itemCount, maxCount)) {
                    uris.add(clipData.getItemAt(i).uri)
                }
            } else if (data.data != null) {
                uris.add(data.data!!)
            }

            if (uris.isEmpty()) {
                promise.reject("MEDIA_ERROR", "No images selected")
                clearPending()
                return
            }

            if (multiple && uris.size > 1) {
                val results = JSONArray()
                for (uri in uris) {
                    val bitmap = decodeBitmapFromUri(uri) ?: continue
                    val resized = resizeIfNeeded(bitmap, maxDimension)
                    val resultStr = buildMediaResult(resized, quality, includeBase64, "pick")
                    results.put(JSONObject(resultStr))
                }
                promise.resolve(results.toString())
            } else {
                val bitmap = decodeBitmapFromUri(uris[0])
                if (bitmap == null) {
                    promise.reject("MEDIA_ERROR", "Failed to decode selected image")
                    clearPending()
                    return
                }
                val resized = resizeIfNeeded(bitmap, maxDimension)
                val result = buildMediaResult(resized, quality, includeBase64, "pick")
                promise.resolve(result)
            }
        } catch (e: Exception) {
            promise.reject("MEDIA_ERROR", "Failed to process library result: ${e.message}", e)
        }

        clearPending()
    }

    // -----------------------------------------------------------------------
    // Helpers
    // -----------------------------------------------------------------------

    private fun decodeBitmapFromUri(uri: Uri): Bitmap? {
        return try {
            val stream = reactApplicationContext.contentResolver.openInputStream(uri)
            val bitmap = BitmapFactory.decodeStream(stream)
            stream?.close()
            bitmap
        } catch (e: Exception) {
            null
        }
    }

    private fun resizeIfNeeded(bitmap: Bitmap, maxDimension: Int?): Bitmap {
        if (maxDimension == null) return bitmap
        val maxDim = maxDimension.toFloat()
        if (bitmap.width <= maxDim && bitmap.height <= maxDim) return bitmap

        val ratio = minOf(maxDim / bitmap.width, maxDim / bitmap.height)
        val newWidth = (bitmap.width * ratio).toInt()
        val newHeight = (bitmap.height * ratio).toInt()

        return Bitmap.createScaledBitmap(bitmap, newWidth, newHeight, true)
    }

    private fun buildMediaResult(bitmap: Bitmap, quality: Int, includeBase64: Boolean, source: String): String {
        val fileName = "${source}_${System.currentTimeMillis()}.jpg"
        val tempFile = File(reactApplicationContext.cacheDir, fileName)

        val stream = ByteArrayOutputStream()
        bitmap.compress(Bitmap.CompressFormat.JPEG, quality, stream)
        val jpegBytes = stream.toByteArray()

        FileOutputStream(tempFile).use { it.write(jpegBytes) }

        val result = JSONObject()
        result.put("uri", "file://${tempFile.absolutePath}")
        result.put("fileName", fileName)
        result.put("mimeType", "image/jpeg")
        result.put("width", bitmap.width)
        result.put("height", bitmap.height)
        result.put("fileSize", jpegBytes.size)
        result.put("timestamp", System.currentTimeMillis())

        if (includeBase64) {
            result.put("base64", Base64.encodeToString(jpegBytes, Base64.NO_WRAP))
        }

        return result.toString()
    }

    private fun clearPending() {
        pendingPromise = null
        pendingOptions = null
        captureUri = null
    }
}

File: native/ios/MediaModule.swift
/**
 * MediaModule - iOS TurboModule for camera capture and photo library access
 *
 * Provides:
 * - captureImage: Opens UIImagePickerController in camera mode
 * - pickFromLibrary: Opens PHPickerViewController for photo library selection
 * - captureFromView: Snapshots an inline SDKCameraView by its cameraId
 * - Permission checking/requesting for camera and photo library
 *
 * Options and results are JSON-stringified for Codegen compatibility.
 * Matches TurboModule spec: NativeMediaModule.ts
 * Registered name: "MediaModule"
 */

import Foundation
import UIKit
import Photos
import PhotosUI
import AVFoundation
import UniformTypeIdentifiers

@objc(MediaModule)
class MediaModule: NSObject {

  // MARK: - Module Registration

  @objc
  static func moduleName() -> String {
    return "MediaModule"
  }

  @objc
  static func requiresMainQueueSetup() -> Bool {
    return true
  }

  // MARK: - Pending promise storage (for delegate callbacks)

  private var pendingResolve: RCTPromiseResolveBlock?
  private var pendingReject: RCTPromiseRejectBlock?
  private var pendingOptions: [String: Any] = [:]

  // MARK: - Capture Image (System Camera)

  @objc
  func captureImage(
    _ optionsStr: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    guard UIImagePickerController.isSourceTypeAvailable(.camera) else {
      reject("MEDIA_ERROR", "Camera is not available on this device", nil)
      return
    }

    guard let options = parseOptions(optionsStr) else {
      reject("MEDIA_ERROR", "Invalid options JSON", nil)
      return
    }

    self.pendingResolve = resolve
    self.pendingReject = reject
    self.pendingOptions = options

    DispatchQueue.main.async {
      let picker = UIImagePickerController()
      picker.sourceType = .camera
      picker.delegate = self
      picker.allowsEditing = false

      // Apply MIME filter
      if let accept = options["accept"] as? String, accept.contains("video") {
        picker.mediaTypes = ["public.image", "public.movie"]
      } else {
        picker.mediaTypes = ["public.image"]
      }

      guard let rootVC = UIApplication.shared.connectedScenes.compactMap({ $0 as? UIWindowScene }).first?.windows.first(where: { $0.isKeyWindow })?.rootViewController else {
        reject("MEDIA_ERROR", "No root view controller available", nil)
        return
      }

      rootVC.present(picker, animated: true)
    }
  }

  // MARK: - Pick From Library (PHPicker)

  @objc
  func pickFromLibrary(
    _ optionsStr: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    guard let options = parseOptions(optionsStr) else {
      reject("MEDIA_ERROR", "Invalid options JSON", nil)
      return
    }

    self.pendingResolve = resolve
    self.pendingReject = reject
    self.pendingOptions = options

    DispatchQueue.main.async {
      if #available(iOS 14.0, *) {
        var config = PHPickerConfiguration()
        let multiple = options["multiple"] as? Bool ?? false
        let maxCount = options["maxCount"] as? Int ?? 10
        config.selectionLimit = multiple ? maxCount : 1
        config.filter = .images

        let picker = PHPickerViewController(configuration: config)
        picker.delegate = self

        guard let rootVC = UIApplication.shared.connectedScenes.compactMap({ $0 as? UIWindowScene }).first?.windows.first(where: { $0.isKeyWindow })?.rootViewController else {
          reject("MEDIA_ERROR", "No root view controller available", nil)
          return
        }

        rootVC.present(picker, animated: true)
      } else {
        // Fallback to UIImagePickerController for iOS < 14
        let picker = UIImagePickerController()
        picker.sourceType = .photoLibrary
        picker.delegate = self
        picker.allowsEditing = false

        guard let rootVC = UIApplication.shared.connectedScenes.compactMap({ $0 as? UIWindowScene }).first?.windows.first(where: { $0.isKeyWindow })?.rootViewController else {
          reject("MEDIA_ERROR", "No root view controller available", nil)
          return
        }

        rootVC.present(picker, animated: true)
      }
    }
  }

  // MARK: - Capture From View

  @objc
  func captureFromView(
    _ cameraId: String,
    options optionsStr: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    guard !cameraId.isEmpty else {
      reject("MEDIA_ERROR", "captureFromView requires a non-empty cameraId", nil)
      return
    }

    guard let options = parseOptions(optionsStr) else {
      reject("MEDIA_ERROR", "Invalid options JSON", nil)
      return
    }

    // Look up the registered camera view by ID and capture its current frame
    DispatchQueue.main.async {
      guard let cameraView = CameraViewManager.shared.getView(byId: cameraId) else {
        reject("MEDIA_ERROR", "Camera view with id '\(cameraId)' not found", nil)
        return
      }

      cameraView.captureFrame { image, error in
        if let error = error {
          reject("MEDIA_ERROR", "Failed to capture frame: \(error.localizedDescription)", error)
          return
        }

        guard let image = image else {
          reject("MEDIA_ERROR", "Captured frame was nil", nil)
          return
        }

        let quality = options["quality"] as? Double ?? 0.8
        // NSNumber from JSONSerialization may not cast directly to Int — use intValue
        let maxDimension = (options["maxDimension"] as? NSNumber)?.intValue
        let includeBase64 = (options["includeBase64"] as? NSNumber)?.boolValue ?? options["includeBase64"] as? Bool ?? false

        let processedImage = self.resizeIfNeeded(image, maxDimension: maxDimension)
        let result = self.buildMediaResult(from: processedImage, quality: quality, includeBase64: includeBase64, source: "viewCapture")
        resolve(result)
      }
    }
  }

  // MARK: - Permissions

  @objc
  func checkCameraPermission(
    _ resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    let status = AVCaptureDevice.authorizationStatus(for: .video)
    resolve(permissionString(from: status))
  }

  @objc
  func checkLibraryPermission(
    _ resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    let status = PHPhotoLibrary.authorizationStatus()
    resolve(photoPermissionString(from: status))
  }

  @objc
  func requestCameraPermission(
    _ resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    AVCaptureDevice.requestAccess(for: .video) { granted in
      resolve(granted ? "granted" : "denied")
    }
  }

  @objc
  func requestLibraryPermission(
    _ resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    if #available(iOS 14.0, *) {
      PHPhotoLibrary.requestAuthorization(for: .readWrite) { status in
        resolve(self.photoPermissionString(from: status))
      }
    } else {
      PHPhotoLibrary.requestAuthorization { status in
        resolve(self.photoPermissionString(from: status))
      }
    }
  }

  // MARK: - Helpers

  private func parseOptions(_ json: String) -> [String: Any]? {
    guard let data = json.data(using: .utf8),
          let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
      return nil
    }
    return obj
  }

  private func permissionString(from status: AVAuthorizationStatus) -> String {
    switch status {
    case .authorized: return "granted"
    case .denied, .restricted: return "denied"
    case .notDetermined: return "undetermined"
    @unknown default: return "undetermined"
    }
  }

  private func photoPermissionString(from status: PHAuthorizationStatus) -> String {
    switch status {
    case .authorized, .limited: return "granted"
    case .denied, .restricted: return "denied"
    case .notDetermined: return "undetermined"
    @unknown default: return "undetermined"
    }
  }

  private func resizeIfNeeded(_ image: UIImage, maxDimension: Int?) -> UIImage {
    guard let maxDim = maxDimension else { return image }
    let maxCGFloat = CGFloat(maxDim)
    let size = image.size

    if size.width <= maxCGFloat && size.height <= maxCGFloat {
      return image
    }

    let ratio = min(maxCGFloat / size.width, maxCGFloat / size.height)
    let newSize = CGSize(width: size.width * ratio, height: size.height * ratio)

    UIGraphicsBeginImageContextWithOptions(newSize, false, 1.0)
    image.draw(in: CGRect(origin: .zero, size: newSize))
    let resized = UIGraphicsGetImageFromCurrentImageContext()
    UIGraphicsEndImageContext()

    return resized ?? image
  }

  fileprivate func buildMediaResult(from image: UIImage, quality: Double, includeBase64: Bool, source: String) -> String {
    let jpegData = image.jpegData(compressionQuality: CGFloat(quality))
    let fileName = "\(source)_\(Int(Date().timeIntervalSince1970 * 1000)).jpg"

    // Save to temp directory
    let tempDir = FileManager.default.temporaryDirectory
    let fileURL = tempDir.appendingPathComponent(fileName)
    try? jpegData?.write(to: fileURL)

    var result: [String: Any] = [
      "uri": fileURL.absoluteString,
      "fileName": fileName,
      "mimeType": "image/jpeg",
      "width": Int(image.size.width),
      "height": Int(image.size.height),
      "fileSize": jpegData?.count ?? 0,
      "timestamp": Int(Date().timeIntervalSince1970 * 1000),
    ]

    if includeBase64, let data = jpegData {
      result["base64"] = data.base64EncodedString()
    }

    guard let jsonData = try? JSONSerialization.data(withJSONObject: result),
          let jsonStr = String(data: jsonData, encoding: .utf8) else {
      return "{}"
    }

    return jsonStr
  }
}

// MARK: - UIImagePickerControllerDelegate

extension MediaModule: UIImagePickerControllerDelegate, UINavigationControllerDelegate {
  func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey: Any]) {
    picker.dismiss(animated: true)

    guard let image = info[.originalImage] as? UIImage else {
      pendingReject?("MEDIA_ERROR", "Failed to get image from picker", nil)
      clearPending()
      return
    }

    let quality = pendingOptions["quality"] as? Double ?? 0.8
    let maxDimension = (pendingOptions["maxDimension"] as? NSNumber)?.intValue
    let includeBase64 = (pendingOptions["includeBase64"] as? NSNumber)?.boolValue ?? false

    let processedImage = resizeIfNeeded(image, maxDimension: maxDimension)
    let result = buildMediaResult(from: processedImage, quality: quality, includeBase64: includeBase64, source: "capture")

    pendingResolve?(result)
    clearPending()
  }

  func imagePickerControllerDidCancel(_ picker: UIImagePickerController) {
    picker.dismiss(animated: true)
    pendingReject?("MEDIA_CANCELLED", "User cancelled media selection", nil)
    clearPending()
  }

  private func clearPending() {
    pendingResolve = nil
    pendingReject = nil
    pendingOptions = [:]
  }
}

// MARK: - PHPickerViewControllerDelegate (iOS 14+)

@available(iOS 14.0, *)
extension MediaModule: PHPickerViewControllerDelegate {
  func picker(_ picker: PHPickerViewController, didFinishPicking results: [PHPickerResult]) {
    picker.dismiss(animated: true)

    if results.isEmpty {
      pendingReject?("MEDIA_CANCELLED", "User cancelled media selection", nil)
      clearPending()
      return
    }

    let quality = pendingOptions["quality"] as? Double ?? 0.8
    let maxDimension = (pendingOptions["maxDimension"] as? NSNumber)?.intValue
    let includeBase64 = (pendingOptions["includeBase64"] as? NSNumber)?.boolValue ?? false
    let multiple = pendingOptions["multiple"] as? Bool ?? false

    let group = DispatchGroup()
    var mediaResults: [String] = []

    for result in results {
      group.enter()
      result.itemProvider.loadObject(ofClass: UIImage.self) { [weak self] object, error in
        defer { group.leave() }
        guard let self = self, let image = object as? UIImage else { return }

        let processedImage = self.resizeIfNeeded(image, maxDimension: maxDimension)
        let resultStr = self.buildMediaResult(from: processedImage, quality: quality, includeBase64: includeBase64, source: "pick")
        mediaResults.append(resultStr)
      }
    }

    group.notify(queue: .main) { [weak self] in
      if multiple && mediaResults.count > 1 {
        // Parse individual results and build array
        var parsed: [[String: Any]] = []
        for r in mediaResults {
          if let data = r.data(using: .utf8),
             let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            parsed.append(obj)
          }
        }
        if let jsonData = try? JSONSerialization.data(withJSONObject: parsed),
           let jsonStr = String(data: jsonData, encoding: .utf8) {
          self?.pendingResolve?(jsonStr)
        } else {
          self?.pendingReject?("MEDIA_ERROR", "Failed to serialize multiple results", nil)
        }
      } else if let first = mediaResults.first {
        self?.pendingResolve?(first)
      } else {
        self?.pendingReject?("MEDIA_ERROR", "No images were loaded from selection", nil)
      }
      self?.clearPending()
    }
  }
}
